File: navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/SessionState.java
package com.mapbox.services.android.navigation.v5.navigation;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.google.auto.value.AutoValue;
import com.mapbox.services.api.directions.v5.models.DirectionsRoute;
import com.mapbox.services.api.directions.v5.models.RouteLeg;

import java.util.Date;

@AutoValue
abstract class SessionState {

  String originalGeometry() {
    return originalDirectionRoute().getGeometry();
  }

  int originalDuration() {
    return (int) originalDirectionRoute().getDuration();
  }

  int originalDistance() {
    return (int) originalDirectionRoute().getDistance();
  }

  int originalStepCount() {
    int stepCount = 0;
    for (RouteLeg leg : originalDirectionRoute().getLegs()) {
      stepCount += leg.getSteps().size();
    }
    return stepCount;
  }

  int currentStepCount() {
    int stepCount = 0;
    for (RouteLeg leg : currentDirectionRoute().getLegs()) {
      stepCount += leg.getSteps().size();
    }
    return stepCount;
  }

  abstract String sessionIdentifier();

  @Nullable
  abstract String originalRequestIdentifier();

  @Nullable
  abstract String requestIdentifier();

  abstract DirectionsRoute originalDirectionRoute();

  abstract DirectionsRoute currentDirectionRoute();

  abstract Date startTimestamp();

  @Nullable
  abstract Date arrivalTimestamp();

  abstract boolean mockLocation();

  abstract int rerouteCount();

  abstract Builder toBuilder();

  static Builder builder() {
    return new AutoValue_SessionState.Builder();
  }


  @AutoValue.Builder
  abstract static class Builder {

    abstract Builder originalDirectionRoute(@NonNull DirectionsRoute directionsRoute);

    abstract Builder currentDirectionRoute(@NonNull DirectionsRoute directionsRoute);

    abstract Builder sessionIdentifier(@NonNull String sessionIdentifier);

    abstract Builder originalRequestIdentifier(@Nullable String originalRequestIdentifier);

    abstract Builder requestIdentifier(@Nullable String requestIdentifier);

    abstract Builder mockLocation(boolean mockLocation);

    abstract Builder rerouteCount(int rerouteCount);

    abstract Builder startTimestamp(@NonNull Date startTimeStamp);

    abstract Builder arrivalTimestamp(@Nullable Date arrivalTimestamp);

    abstract SessionState build();

  }
}

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/events/SessionState.kt
package com.mapbox.navigation.core.telemetry.events

import android.location.Location
import com.mapbox.api.directions.v5.models.DirectionsRoute
import com.mapbox.navigation.core.telemetry.obtainGeometry
import com.mapbox.navigation.core.telemetry.obtainStepCount
import java.util.Date

internal data class SessionState(
    var secondsSinceLastReroute: Int = -1,
    var eventRouteProgress: MetricsRouteProgress = MetricsRouteProgress(null),
    var eventLocation: Location = Location(MetricsLocation.PROVIDER).apply {
        latitude = 0.0
        longitude = 0.0
    },
    var eventDate: Date? = null,
    var eventRouteDistanceCompleted: Double = 0.0,
    var afterEventLocations: List<Location>? = null,
    var beforeEventLocations: List<Location>? = null,
    var originalDirectionRoute: DirectionsRoute? = null,
    var currentDirectionRoute: DirectionsRoute? = null,
    var sessionIdentifier: String = "",
    var tripIdentifier: String = "",
    var originalRequestIdentifier: String? = null,
    var requestIdentifier: String? = null,
    var mockLocation: Boolean = false,
    var rerouteCount: Int = 0,
    var startTimestamp: Date? = null,
    var arrivalTimestamp: Date? = null,
    var locationEngineName: String = "",
    var percentInForeground: Int = 100,
    var percentInPortrait: Int = 100
) {

    // TODO:OZ Clean up never used functions below
    //  We should also review SessionState properties as some of them are set but never used
    //  E.g. percentInForeground and percentInPortrait set twice
    //  see MapboxNavigationTelemetry#populateSessionState and MapboxNavigationTelemetry#populateNavigationEvent
    //  and never referenced in RerouteEvent / NavigationRerouteEvent
    /*
    * Original route values
    */
    fun originalStepCount(): Int = obtainStepCount(originalDirectionRoute)

    fun originalGeometry(): String = obtainGeometry(originalDirectionRoute)

    fun originalDistance(): Int = originalDirectionRoute?.distance()?.toInt() ?: 0

    fun originalDuration(): Int = originalDirectionRoute?.duration()?.toInt() ?: 0

    /*
    * Current route values
    */
    fun currentStepCount(): Int = obtainStepCount(currentDirectionRoute)

    fun currentGeometry(): String = obtainGeometry(currentDirectionRoute)
}
